### PR TITLE
Avoid `PreparedStatement`s in the `JdbcColumnType` API

### PR DESCRIPTION
**Pull request by dmytro-grankin**
Until now, column values in `JdbcColumnType` were set using `PreparedStatement`. This approach exposed implementation details, so the changes in the implementation would cause changes in the API.

This PR introduces `Parameters` class, which serves as a DTO for SQL query parameters. It allows to hide the details about construction of queries in a place, where query parameters are specified. From now, `JdbcColumnType` uses `Parameteres`, which solves the described issue.

Also `StorageQuery` descendants and `IdColumn` were reworked and now use `Parameters`. So setting of parameters to a `PreparedStatement` is placed in `StorageQuery` only. Such changes bring good generalization and make future changes easier.

Previously, to set parameters for `PreparedStatement`, different methods of `PreparedStatement` were used (e.g. `setString(...)`, `setInt(...)`, `setBytes(...)`). To make possible setting of `PreparedStatement` parameters in a single place, `setObject(...)` method is used now. To guarantee correct work with different types, `java.sql.Types` are used to specify a type of a parameter.

The version was increased to `0.9.77-SNAPSHOT`.

### CODE REVIEW

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/IdColumn.java [hunk 1]
```diff
@@ -83,7 +75,7 @@
     }
 
     public static IdColumn<String> typeString(String columnName) {
```
Comment (armiol) [inline rdbms/src/main/java/io/spine/server/storage/jdbc/IdColumn.java]: Please have it documented as well.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+import io.spine.server.storage.jdbc.Sql;
+import io.spine.server.storage.jdbc.TableColumn;
+
+import javax.annotation.Nullable;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A parameter of a SQL query.
```
Comment (armiol) [inline rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java]: `of an SQL query`.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * <p>Contains all the necessary data to work with its raw value.
```
Comment (armiol) [inline rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java]: This sentence isn't specific enough. Let's discuss what it should be.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    @Nullable
+    public Object getValue() {
+        return value;
+    }
+
+    /**
+     * Obtains a SQL type of the parameter.
+     *
+     * <p>A type allows to store a raw value and work with objects of different classes
```
Comment (armiol) [inline rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java]: I don't understand, how type allows to store values.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A parameter of a SQL query.
+ *
+ * <p>Contains all the necessary data to work with its raw value.
+ *
+ * @author Dmytro Grankin
+ */
+public class Parameter {
```
Comment (armiol) [inline rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java]: It looks `final`. Same as `Parameters`.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        this.parameters = parameters;
+    }
+
+    /**
+     * Obtains a set of parameter identifiers.
+     *
+     * @return parameter identifiers
+     */
+    public Set<Integer> getIdentifiers() {
+        return parameters.keySet();
```
Comment (armiol) [inline rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java]: That looks suspicious. I wouldn't return a collection like this.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * <p>This class allows to encapsulate setting of parameters to a concrete implementation.
+ *
+ * @author Dmytro Grankin
+ */
+public class Parameters {
+
+    private final Map<Integer, Parameter> parameters;
+
+    private Parameters(Map<Integer, Parameter> parameters) {
```
Comment (armiol) [inline rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java]: This isn't how a `Builder` works. Please check similar examples.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+     * @return the query parameter
+     * @throws IllegalArgumentException if there is no parameters with the specified identifier
+     */
+    public Parameter getParameter(Integer identifier) {
+        checkArgument(parameters.containsKey(identifier));
+        final Parameter value = parameters.get(identifier);
+        return value;
+    }
+
+    public static Parameters empty() {
```
Comment (armiol) [inline rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java]: Please document the public API.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/IdColumn.java (hunk 1 above)
Comment (dmytro-grankin) [reply rdbms/src/main/java/io/spine/server/storage/jdbc/IdColumn.java]: Done.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/IdColumn.java [hunk 9]
```diff
@@ -101,20 +93,32 @@ protected IdColumn(String columnName) {
      */
     public abstract Class<I> getJavaType();
 
+    /**
+     * Normalizes the identifier before setting it to a {@link PreparedStatement}.
```
[the remark this answers is not in the document]
Comment (dmytro-grankin) [reply rdbms/src/main/java/io/spine/server/storage/jdbc/IdColumn.java]: Done.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java (hunk 2 above)
Comment (dmytro-grankin) [reply rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java]: Done.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java (hunk 5 above)
Comment (dmytro-grankin) [reply rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java]: Done.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java (hunk 7 above)
Comment (dmytro-grankin) [reply rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java]: Done.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java (hunk 6 above)
Comment (dmytro-grankin) [reply rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java]: Done.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java (hunk 8 above)
Comment (dmytro-grankin) [reply rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameters.java]: Done.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java (hunk 4 above)
Comment (dmytro-grankin) [reply rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java]: Done.

File: rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java (hunk 3 above)
Comment (dmytro-grankin) [reply rdbms/src/main/java/io/spine/server/storage/jdbc/query/Parameter.java]: Done.

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

**Pull request merged**